### ISSUE TITLE
FTP plugin usability and SCP/SFTP mode issues"

### ISSUE THREAD

**Issue by freebrowser1** [doublecmd/doublecmd#1604]
### Version

Double commander Version 1.1.13 gamma
Revision 453
Commit f60bf09df
Build 2024/04/20
Lazarus 3.99.0.0
Free Pascal 3.2.2
aarch64-Darwin-cocoa
Mac OS X 14.2.0

### Description

When attempting using SFTP, I got several issues which other file browsers (like Forklift or even the terminal Midnight Commander) do not have.

1. (See #227) libssh2 is missing. The path to libssh2 appears to be hardcoded to /usr/local/lib/ which is standard on Homebrew on Intel Macs (where it works), but not on M1/2/3/4 (arm64) Macs. I created a workaround by symlinking this by `ln -s /opt/homebrew/lib/libssh2.dylib /usr/local/lib/libssh2.dylib` and then ssh 'works' but with lots of limitations.

2. .local domains appear to be unsupported, which means that local network devices in the same wired or wifi LAN can only be approached by hard IP addresses which may change. Other apps on the same Mac do recognize .local domains.

3. Connection to Termux SSH server (Termux is a terminal app for Android with SSH server with port 8022) I even get SSH ERROR without any explanation. Other SSH clients connect normally with [ipaddress]:8022.
<img width="434" alt="Screenshot 2024-05-03 at 12 42 28" src="https://github.com/doublecmd/doublecmd/assets/26625141/d424ece8-ebec-4c98-a235-d6ca098b084d">

```
Info: [3], Connecting to: 192.168.0.7
Info: [1], 
Info: [3], Connection established
Info: [3], Key exchange method: curve25519-sha256
Info: [3], Encryption method: aes256-gcm@openssh.com
Info: [3], Host key method: ecdsa-sha2-nistp256
Info: [3], Server fingerprint: (MD5) 73 D3 3C DB B0 EB AF 13 FB 70 20 B2 82 EA D1 8C
Info: [3], Server fingerprint: (SHA1) 9A 85 A3 D1 4E 04 5D 0A 50 09 34 1F 3C CC 05 9F 4A C9 83 F9
Info: [3], Server fingerprint: (SHA256) oyGjb9fkzeCgzIPciHZHtFev+M0vb4c3xj6tkt8KID8
Info: [3], Authentication methods: publickey,keyboard-interactive
Info: [3], SSH-agent authentication
Error: [6], SSH ERROR 
```

4. When filled in ~ in the 'Remote dir' field, it switched to the root folder. Only hard paths are allowed.

**Comment by alexx2000** [comment]
1. It is not harcoded, `/usr/local/lib` is a standard system library path. Current version should work with `/opt/homebrew/lib` without additional symlinks.

**Comment by lprot** [comment]
Tested SSH on:
Double Commander
Version: 1.2.0 alpha
Revision: 11320
Commit: 65e72c0d3
Build date: 2024/05/09
Lazarus: 3.99.0.0
FPC: 3.2.2
Platform: x86_64-Darwin-cocoa
OS version: Mac OS X 12.7.4

What is immediately user unfriendly here that the popup has alot fo text but does not give a hint how to really install the library:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/3f05a9ff-3bf3-4385-9847-70ac1c78f7aa)
Why not to change this popup to simply:
Title: LibSSH2 > "LibSSH2 not found!"
Popup text: 
To install:
1. Install https://brew.sh/
2. Run Terminal and enter: brew install libssh2
3. Restart Double Commander

**Comment by lprot** [comment]
To test SSH connection you can register at sdf.org
![image](https://github.com/doublecmd/doublecmd/assets/2608545/848941d6-01fc-482e-85c2-b56a209f19e3)
And yes, DC cannot connect via SSH to the server but in zsh I had no problems with that.

**Comment by alexx2000** [comment]
@lprot this server does not support SFTP, it has only SSH (only shell).

> Why not to change this popup to simply:

Because DC is a cross-platform application, not macOS only.

**Comment by lprot** [comment]
> @lprot this server does not support SFTP, it has only SSH (only shell).
> 
> > Why not to change this popup to simply:
> 
> Because DC is a cross-platform application, not macOS only.

Then makes sense to add a link to this popup to the help page where it is shortly explained per platform how to install this lib :)

**Comment by lprot** [comment]
> @lprot this server does not support SFTP, it has only SSH (only shell).

So DC does not support SCP?

**Comment by alexx2000** [comment]
SSH is a low level protocol, SFTP is a high level protocol which works over SSH.

Also DC supports SSH+SCP, it can be enabled at the `Advanced` page. That's why this option called `SSH`. 

But SCP is a `secure copy` only. Other actions is done using console commands (`ls mkdir rmdir` etc) which may not work with some servers.

FTP is a File Transfer Plugin here.

**Comment by lprot** [comment]
> SSH is a low level protocol, SFTP is a high level protocol which works over SSH.
> 
> Also DC supports SSH+SCP, it can be enabled at the `Advanced` page. That's why this option called `SSH`.
> 
> But SCP is a `secure copy` only. Other actions is done using console commands (`ls mkdir rmdir` etc) which may not work with some servers.
> 
> FTP is a File Transfer Plugin here.

It is not user friendly now. Why not to make instead of TLS SSH and SCP in advanced, a radio button with TLS, SFTP, SCP? That would be much more intuitive. Tried SCP and it does not work as well:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/702d485b-9cb9-420d-9f58-b6f4c2da8b17)

**Comment by alexx2000** [comment]
I fixed a crash. But this server is very non-standard so plugin can not work with it.

**Comment by freebrowser1** [comment]
> I fixed a crash. But this server is very non-standard so plugin can not work with it.
I agree with 'lprot' that the user interface is not very intuitive. 
Any other SSH / SFTP file browser works flawlessly on macOS.

**Comment by lprot** [comment]
> I agree with 'lprot' that the user interface is not very intuitive.

I beleive we just need to remove here "TLS" and "SSH" checkboxes, remove "Use SCP+SSH protocol (no SFTP)" from "Advanced" and add instead of all this "Mode:" listbox with "FTP, SFTP(TLS), SFTP, SCP". This listbox should be above the "Host:[Port:]:" inputbox. This would make UI much more intuitive.

**Comment by lprot** [comment]
> I fixed a crash. But this server is very non-standard so plugin can not work with it.

I beleive the only speciality is that that server asks about pressing Enter 3 times and only after that lets you to do ls.
It makes sense to make 3 tries with sending Enter press and check if ls or whatever command answers with the listing. And only then give up. That will let to make the connection more robust as servers may different banners setup etc.

**Comment by lprot** [comment]
@alexx2000 I can confirm that current SSH plugin does not work with Termux. To reproduce:
1. On the smartphone in browser visit https://github.com/termux/termux-app/releases and download and install APK of release 0.118.0
2. Run Termux and install sshd with: pkg install openssh
3. Set password via: passwd
4. Run SSH server: sshd

Port is always 8022 and WiFi IP of my smartphone is 192.168.0.53. So I connected like follows:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/444b08cc-0bbe-4f25-8a7e-65e1636b370d)
![image](https://github.com/doublecmd/doublecmd/assets/2608545/5a043dd1-f3e5-4298-838d-8c52c8487876)
Got error:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/a040fbe1-6b90-4c21-981e-ff00e10481e2)

Tried to connect differently:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/cde66337-6d2d-46f9-9f2b-36eaf7fbf382)
![image](https://github.com/doublecmd/doublecmd/assets/2608545/5d5932d6-3733-4c4b-98a7-72eb53cbc3bd)

Got:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/d7105479-808c-4e05-a4b2-d56e291bce3a)

ZSH has no problems with connection:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/c5227854-5c1c-42d9-8eec-63c1ee0bc1d6)

**Comment by lprot** [comment]
The problem here is definately with SCP mode and with ls logic of FTP plugin. If we take a look at SDF.org and at Termux shells, we can see that the only proper way is to use ls -al and parse the text from "total " until but not including the last line:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/3878bdac-ea24-4b7c-a5de-ffce07352670)

**Comment by freebrowser1** [comment]
It still is weird. Any other SSH / SFTP client connects properly with a Termux server on its standard port 8022 and can transfer files.
The only issue is, is that the root folder / is not accessible by the 'Termux user' but that is also the case when one logs in on any linux or u**x system, including macOS itself as a non-root user.
When filling in `/data/data/com.termux/files/home` (which is the home folder) as the startup dir when connecting to Termux server, it refuses to connect either. 
It is definely an issue in DC.

**Comment by alexx2000** [comment]
Termux uses a standard OpenSSH server. I tested it, DC works with it without any problems.

**Comment by lprot** [comment]
> Termux uses a standard OpenSSH server. I tested it, DC works with it without any problems.

As we see from the logs and screenshots above, current ftp plugin does not work with Termux on Android neither in SFTP nor in SCP mode.

**Comment by alexx2000** [comment]
It works fine with my phone (Termux and OpenSSH) in the both modes.

**Comment by lprot** [comment]
> It works fine with my phone (Termux and OpenSSH) in the both modes.

How connection settings look like?

**Comment by alexx2000** [comment]
Everything is like yours.

**Comment by freebrowser1** [comment]
DC has now completely stopped working with SFTP.
<img width="453" alt="Screenshot 2024-05-15 at 09 18 54" src="https://github.com/doublecmd/doublecmd/assets/26625141/ed25b1ea-1c03-4ffd-b34f-eb6c93e0afd5">

"Cannot connect to server"

When I replace "nothingphone.local" by the hard internap IP DC locks up and I have to Force Quit it.
The domain name "nothingphone.local" can normally be reached and connected by other SSH/SFTP clients from my Mac.

**Comment by alexx2000** [comment]
Try with latest snapshot, maybe it will write a more details in the log.

**Comment by lprot** [comment]
Double Commander
Version: 1.2.0 alpha
Revision: 11324
Commit: d2416cc2a
Build date: 2024/05/19
Lazarus: 3.99.0.0
FPC: 3.2.2
Platform: x86_64-Darwin-cocoa
OS version: Mac OS X 12.7.5

I can confirm that I can connect to the Termux via SFTP:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/dc5d602e-bfa2-403c-9483-10aced8074de)

And via SCP as well:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/d84f0b35-84f9-4cd3-90cc-dba93277ebca)

There is a strange stat bug in SFTP mode.

**Comment by lprot** [comment]
> DC has now completely stopped working with SFTP. 
> 
> "Cannot connect to server"
> 
> When I replace "nothingphone.local" by the hard internap IP DC locks up and I have to Force Quit it. The domain name "nothingphone.local" can normally be reached and connected by other SSH/SFTP clients from my Mac.

@alexx2000 Can cofirm it https://github.com/doublecmd/doublecmd/issues/1654

**Comment by lprot** [comment]
> DC has now completely stopped working with SFTP. <img alt="Screenshot 2024-05-15 at 09 18 54" width="453" src="https://private-user-images.githubusercontent.com/26625141/330698816-ed25b1ea-1c03-4ffd-b34f-eb6c93e0afd5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYxODcwMjQsIm5iZiI6MTcxNjE4NjcyNCwicGF0aCI6Ii8yNjYyNTE0MS8zMzA2OTg4MTYtZWQyNWIxZWEtMWMwMy00ZmZkLWIzNGYtZWI2YzkzZTBhZmQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIwVDA2MzIwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlNWQzZDgxMDQyNjI0M2JhNzU4Nzg0NzUyODlmNmFiNTZmYmJjYmFiMGZhYzM0NjNiYjE5ZDg5OTNlZjc1MmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.d7-y_jh_EM-1cq9L6DGn7fSytwy5xLZmnCYYziRtu2Q">
> 
> "Cannot connect to server"
> 
> When I replace "nothingphone.local" by the hard internap IP DC locks up and I have to Force Quit it. The domain name "nothingphone.local" can normally be reached and connected by other SSH/SFTP clients from my Mac.

Via "Quick connect" but it works via "Add connection".

**Comment by lprot** [comment]
@freebrowser1 I see no bugs listed in this issues in 
Double Commander
Version: 1.2.0 alpha
Revision: 11332
Commit: afc3a9fc9
Build date: 2024/05/20
Lazarus: 3.99.0.0
FPC: 3.2.2
Platform: x86_64-Darwin-cocoa
OS version: Mac OS X 12.7.5

And the new connection interface is implemented https://github.com/doublecmd/doublecmd/issues/1654

You can close the issue as completed?

**Comment by freebrowser1** [comment]
Where did you get the version 1.20 ? I see only the latest as 1.114.

**Comment by alexx2000** [comment]
https://doublecmd.sourceforge.io/site/eng/snapshots.php

**Comment by freebrowser1** [comment]
Indeed, Termux does work (when added with 'Add connection') but only with a hard IP address and not with a .local domain.
And private keys don't work neither.
"Error: [6], SSH ERROR Username/PublicKey combination invalid"
Another issue is when I set F4 edit to an external app and set '/Applications/VSCodium.app' it does not work. So custom / external apps for viewing text files does not work.

**Comment by lprot** [comment]
> Indeed, Termux does work (when added with 'Add connection') but only with a hard IP address and not with a .local domain. And private keys don't work neither. "Error: [6], SSH ERROR Username/PublicKey combination invalid" Another issue is when I set F4 edit to an external app and set '/Applications/VSCodium.app' it does not work. So custom / external apps for viewing text files does not work.

Don't know what is different on your side but for me .local works. My settings:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/8eeb3165-5383-433d-aee5-cfa32d2911b8)

**Comment by freebrowser1** [comment]
```
ping xubuntu.local
PING xubuntu.local (192.168.1.145): 56 data bytes
64 bytes from 192.168.1.145: icmp_seq=0 ttl=64 time=10.682 ms
64 bytes from 192.168.1.145: icmp_seq=1 ttl=64 time=6.000 ms
^C
--- xubuntu.local ping statistics ---
2 packets transmitted, 2 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 6.000/8.341/10.682/2.341 ms
```
<img width="449" alt="Screenshot 2024-05-22 at 19 33 10" src="https://github.com/doublecmd/doublecmd/assets/26625141/a34dd9ec-6eb2-42e4-b1f9-fe96dff5227b">

<img width="435" alt="Screenshot 2024-05-22 at 19 33 33" src="https://github.com/doublecmd/doublecmd/assets/26625141/26709d27-9e45-48e1-a0ca-66a0d18f26ea">

<img width="273" alt="Screenshot 2024-05-22 at 19 33 42" src="https://github.com/doublecmd/doublecmd/assets/26625141/f317b39d-7e2d-49a5-84ee-ddfe4de9ca69">

When I replace 'xubuntu.local' with a hard IP address it does work.

**Comment by lprot** [comment]
@freebrowser1 you are lucky that you can connect at all. In my case I cannot connect by using keys.

I configure and run sshd on macOS like this:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/efd5ffb3-0561-42d5-a44a-93a528fe14cd)

Then configure FTP plugin like this:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/c778c3c7-4e55-4141-a5fc-adf3dd08e547)
![image](https://github.com/doublecmd/doublecmd/assets/2608545/a2b4e769-141c-4fd5-9c8b-923aacfd14dd)

Try to connect like this:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/2dda1ffe-dca8-42b4-9a82-3ba9e90a09b8)

Get the result like this:
![image](https://github.com/doublecmd/doublecmd/assets/2608545/5bafe165-1c21-41b4-b2c4-cc522a8b8172)

@alexx2000 no matter what I try, SFTP or SCP, localhost or 127.0.0.1 result is always the same like above. Without keys I can connect with keys not. I also tried RSA.

**Comment by alexx2000** [comment]
https://support.bdrive.com/t/sftp-connection-failure-username-publickey-combination-invalid/219137

Try to use ed25519 key (console `ssh` uses this key).

**Comment by lprot** [comment]
> https://support.bdrive.com/t/sftp-connection-failure-username-publickey-combination-invalid/219137
> 
> It is a OpenSSH-server misconfiguration. Try to use ed25519 key (console `ssh` client uses this key).

![image](https://github.com/doublecmd/doublecmd/assets/2608545/afb5e1b4-4bfe-4122-8d13-082b5ddc7916)
![image](https://github.com/doublecmd/doublecmd/assets/2608545/48324956-ed15-474a-a165-805c6f492912)
![image](https://github.com/doublecmd/doublecmd/assets/2608545/08e7ce4b-a9ff-432a-b50e-2d540b3a387e)
![image](https://github.com/doublecmd/doublecmd/assets/2608545/7e5ebf36-131c-4491-8475-fdd413750bab)

**Comment by alexx2000** [comment]
It is needed to inspect the SSH-server’s authentication log to see why it gives this answer.

**Comment by lprot** [comment]
> It is needed to inspect the SSH-server’s authentication log to see why it gives this answer.
> 
> Key authentication works fine with all servers which I use.
> 
> Check that your `authorized_keys` file contains a correct key.

@alexx2000 Bad me! I missed to authorize the generated keys with "ssh-copy-id lprot@localhost":
![image](https://github.com/doublecmd/doublecmd/assets/2608545/c8e4fed1-777c-449e-b7bb-41bb3d293547)


After that all works with settings like this:

![image](https://github.com/doublecmd/doublecmd/assets/2608545/2d968a42-2e9f-413c-a7a6-0bb16a4ef3d0)
![image](https://github.com/doublecmd/doublecmd/assets/2608545/db7d9212-906c-4032-9b50-6210b666c03b)
![image](https://github.com/doublecmd/doublecmd/assets/2608545/32f97be1-2e15-4c3e-a697-b161bb7f0952)

@freebrowser1 I beleive we need to close this issue and start to handle separate issues if any remain. I tried 127.0.0.1:22 and localhost:22 and both works.

**Comment by lprot** [comment]
Just for the info: I had a chance to extensively test SFTP and SCP connection to sshd in the latest DC version under Windows and did not have any connection related issues.

**Comment by freebrowser1** [comment]
The version 1.1.18 now works (under macOS 15 Sequoia) but still can't connect to local domains, one has to fill in a hard IP address.

**Comment by alexx2000** [comment]
What is a "local domain"? How can I test it?

**Comment by freebrowser1** [comment]
This is done by mDNS, so local devices in the same LAN have a .local name. This evades using ever changing indoor IP addresses like 192.168.x.y or 10.x.y.z.
https://en.wikipedia.org/wiki/.local

**Comment by alexx2000** [comment]
I guess "local domain" should work now.

**Comment by freebrowser1** [comment]
This version is not yet available for download, it is still 1.1.19 of Oct 6.

**Comment by freebrowser1** [comment]
I tried the snapshot and connected to a name.local domain which is recognized by other apps, but not by this 1.2.0 shapshot.
So it is not fixed yet.

**Comment by JonnyTech** [comment]
In the DC snapshot, please go to: Menu → Help → About, then copy the info here.

**Comment by freebrowser1** [comment]
> In the DC snapshot, please go to: Menu → Help → About, then copy the info here.

Double Commander
Version: 1.2.0 alpha
Revision: 11643
Commit: d498e63c2
Build date: 2024/11/01
Lazarus: 4.99.0.0
FPC: 3.2.2
Platform: aarch64-Darwin-cocoa
OS version: Mac OS X 15.1.0

**Comment by JonnyTech** [comment]
Thanks, ahh macOS user, @Skif-off are you able to check?

**Comment by Skif-off** [comment]
@JonnyTech 
Sorry, I used FTP, including on local networks (192.168.x.y), but I don't understand what "connected to a name.local domain" means.

**Comment by lprot** [comment]
@freebrowser1 how do you configure mDNS on Termux so it is accessible by termux.local ?

**Comment by freebrowser1** [comment]
I use avahi on the Termux server (which is an Android 14 phone).
I tried again and now it did work !
```
Info: [3], Connecting to: nothingphone.local
Info: [3], nothingphone.local:8022
Info: [3], 192.168.0.9:8022
Info: [1], 
Info: [3], Connection established
Info: [3], Key exchange method: curve25519-sha256
Info: [3], Encryption method: aes256-gcm@openssh.com
Info: [3], Host key method: ecdsa-sha2-nistp256
Info: [3], Server fingerprint: (MD5) 18 10 88 C9 A8 37 D0 80 30 9D 82 47 9F FD 8B 01
Info: [3], Server fingerprint: (SHA1) 96 31 D8 F6 5D E4 F2 ED 92 01 C3 EC 3A 7A 70 81 50 36 59 04
Info: [3], Server fingerprint: (SHA256) zHAhhMruHtdidqM1FcVx/YVVaDLtRV6EJVJV1t8qnIY
Info: [3], Authentication methods: publickey,password,keyboard-interactive
Info: [3], Password authentication
Info: [3], Authentication succeeded
Info: [3], Remote directory: /data/data/com.termux/files/home
Info: [1], CONNECT /nothingphone.local          
```
